force-app/main/default/classes/CaseServiceTest.cls

```diff
@@ -0,0 +1,192 @@
+        
+        // When 
+
+        Test.startTest();
+
+        Database.SaveResult result = Database.insert(request, false);
+
+        Test.stopTest();
+
+        // Then
+        System.assert(!result.isSuccess());
+        System.assert(result.getErrors().size() > 0);
+        System.assertEquals('Phone request are only allowed to have three attendee',
+        result.getErrors()[0].getMessage());
+
+    }
+    
+    @isTest
+    public static void bookEventFromTheWeb(){
+        // Given
+        Account a1 = new Account(Name = 'Salesforce');
+        insert a1;
+        Contact c1 = new Contact(LastName = 'Contact 1', AccountId = a1.ID);
+        insert c1;
+
+        Case request = new Case(Origin = 'Web',Subject = 'Event Booking', Description = 'Book One Spot for Company A1', AccountId = a1.ID);
+
+        // When 
+        Test.startTest();
+        Database.SaveResult result = Database.insert(request, false);
+        Test.stopTest();
+
+        // Then
+        System.assert(result.isSuccess());
+    } 
+        
+    @isTest
+    public static void refuseEventBookingWhenNumberOfAttendeesIsMoreThanOneForWebCases(){
+        // Given
+        
+        Account a1 = new Account(Name = 'Salesforce');
+        insert a1;
+        
+        Contact c1 = new Contact(LastName = 'Contact 1', AccountId = a1.ID);
+        insert c1;
+        Contact c2 = new Contact(LastName = 'Contact 2', AccountId = a1.ID);
+        insert c2;
+        Contact c3 = new Contact(LastName = 'Contact 3', AccountId = a1.ID);
+        insert c3;
+        Case request = new Case(Origin = 'Web',Subject = 'Event Booking', Description = 'Book Spots for Company A1 & their 3 employees', AccountId = a1.ID);
+        
+        // When 
+
+        Test.startTest();
+        Database.SaveResult result = Database.insert(request, false);
+        Test.stopTest();
+
+        // Then
+
+        System.assert(!result.isSuccess());
+        System.assert(result.getErrors().size() > 0);
+        System.assertEquals('Web request are only allowed to have one attendee', result.getErrors()[0].getMessage());
+
+            
+    }
+
+    @isTest
+    public static void registerContactsWhenCaseIsClosed() {
+        // Given     
+        Test.setMock(HttpCalloutMock.class, new RegistrationCalloutMock(201));
+        Account a1 = new Account(Name = 'Salesforce');
+        insert a1;
+        Contact c1 = new Contact(LastName = 'Contact 1', AccountId = a1.ID, Email = 'test1@test.com');
+        insert c1;
+        Case request = new Case(Origin = 'Phone',Subject = 'Event Booking', Description = 'simple description', AccountId = a1.ID);
+        insert request;
+
+        // when
+        Test.startTest();
+        request.Status = 'Closed';
+        Database.SaveResult result = Database.update(request, false);
+        Test.stopTest();
+
+        List<Log__c> logs = [SELECT Message__c FROM Log__c];
```

Comment:
Instead of using logs, it's better to use the native Limits.getEmailInvocations() to know if any emails were sent !